File: services/product-service/src/controllers/product.controller.ts
import { Request, Response } from 'express';
import { ProductService } from '../services/product.service';


export class ProductController {
  private service: ProductService;

  constructor() {
    this.service = new ProductService();
  }

  createProduct = async (req: Request, res: Response) => {
    try {
        const product = await this.service.createProduct(req.body);
        res.status(201).json(product);
    } catch (error: any) {
        res.status(400).json({error: error.message})
    }
  }

  getProducts = async (req: Request, res: Response) => {
    try {
      const query = {
        factoryId: req.query.factoryId as string,
        categoryId: req.query.categoryId as string,
        status: req.query.status as any,
        search: req.query.search as string,
        page: req.query.page ? parseInt(req.query.page as string) : 1,
        limit: req.query.limit ? parseInt(req.query.limit as string) : 20
      };
      const result = await this.service.getProducts(query);
      res.json(result);
    } catch (error: any) {
      res.status(400).json({ error: error.message });
    }
  };

  getProductBySlug = async (req: Request, res: Response) => {
    try {
        const product = await this.service.getProductBySlug(req.params.slug);
        res.json(product)
    }catch (error: any){
        res.status(404).json({error: error.message})
    }
  }

  getProductById = async (req: Request, res: Response) => {
    try {
        const product = await this.service.getProductById(req.params.id);
        res.json(product)
    } catch (error: any){
        res.status(404).json({error: error.message})
    }
  }
  getVariantById = async (req: Request, res: Response) => {
    try {
      const variant = await this.service.getVariantById(req.params.variantId);
      res.json({ success: true, data: variant });
    } catch (error: any) {
      res.status(404).json({ success: false, error: error.message });
    }
  };

  updateProduct = async (req: Request, res: Response) => {
    try {
        const product = await this.service.updateProduct(req.params.id, req.body);
        res.json(product)
    } catch (error: any) {
        res.status(400).json({error: error.message})
    }
  }
  publishProduct = async (req: Request, res: Response) => {
    try {
      const product = await this.service.publishProduct(req.params.id);
      res.json(product);
    } catch (error: any) {
      res.status(400).json({ error: error.message });
    }
  };

  deleteProduct = async (req: Request, res: Response) => {
    try {
      await this.service.deleteProduct(req.params.id);
      res.status(204).send();
    } catch (error: any) {
      res.status(400).json({ error: error.message });
    }
  };

  addImages = async (req: Request, res: Response) => {
    try {
      await this.service.addProductImages(req.params.id, req.body.images);
      res.status(201).json({ message: 'Images added successfully' });
    } catch (error: any) {
      res.status(400).json({ error: error.message });
    }
  };

  createVariant = async (req: Request, res: Response) => {
    try {
      const variant = await this.service.createVariant(req.body);
      res.status(201).json(variant);
    } catch (error: any) {
      res.status(400).json({ error: error.message });
    }
  };

  // ============= Grosir Config Management =============

  setGrosirAllocations = async (req: Request, res: Response) => {
    try {
      const { allocations } = req.body;

      if (!allocations || !Array.isArray(allocations) || allocations.length === 0) {
        return res.status(400).json({ error: 'Allocations array is required' });
      }

      const result = await this.service.setGrosirAllocations(req.params.id, allocations);
      res.json({
        success: true,
        message: 'Grosir allocations set successfully',
        data: result
      });
    } catch (error: any) {
      res.status(400).json({ success: false, error: error.message });
    }
  };

  setWarehouseTolerance = async (req: Request, res: Response) => {
    try {
      const { tolerances } = req.body;

      if (!tolerances || !Array.isArray(tolerances) || tolerances.length === 0) {
        return res.status(400).json({ error: 'Tolerances array is required' });
      }

      const result = await this.service.setWarehouseTolerance(req.params.id, tolerances);
      res.json({
        success: true,
        message: 'Warehouse tolerance set successfully',
        data: result
      });
    } catch (error: any) {
      res.status(400).json({ success: false, error: error.message });
    }
  };

  getGrosirConfig = async (req: Request, res: Response) => {
    try {
      const config = await this.service.getGrosirConfig(req.params.id);
      res.json({
        success: true,
        data: config
      });
    } catch (error: any) {
      res.status(404).json({ success: false, error: error.message });
    }
  };
}

File: services/product-service-go/internal/controller/product_controller.go
package controller

import (
	"errors"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/lakoo/product-service-go/internal/service"
	"github.com/lakoo/product-service-go/types"
)

type ProductController struct {
	service *service.ProductService
}

func NewProductController(svc *service.ProductService) *ProductController {
	return &ProductController{service: svc}
}

// POST /api/products
func (c *ProductController) CreateProduct(ctx *gin.Context) {
	var dto types.CreateProductDTO
	if err := ctx.ShouldBindJSON(&dto); err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	product, err := c.service.CreateProduct(ctx.Request.Context(), dto)
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusCreated, product)
}

// GET /api/products
func (c *ProductController) GetProducts(ctx *gin.Context) {
	var query types.ProductQuery
	if err := ctx.ShouldBindQuery(&query); err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	result, err := c.service.GetProducts(ctx.Request.Context(), query)
	if err != nil {
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusOK, result)
}

// GET /api/products/:slug
func (c *ProductController) GetProductBySlug(ctx *gin.Context) {
	slug := ctx.Param("slug")

	product, err := c.service.GetProductBySlug(ctx.Request.Context(), slug)
	if err != nil {
		if errors.Is(err, service.ErrProductNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Product not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusOK, product)
}

// GET /api/products/id/:id
func (c *ProductController) GetProductByID(ctx *gin.Context) {
	idStr := ctx.Param("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: "Invalid product ID"})
		return
	}

	product, err := c.service.GetProductByID(ctx.Request.Context(), id)
	if err != nil {
		if errors.Is(err, service.ErrProductNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Product not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusOK, product)
}

// PATCH /api/products/:id
func (c *ProductController) UpdateProduct(ctx *gin.Context) {
	idStr := ctx.Param("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: "Invalid product ID"})
		return
	}

	var dto types.UpdateProductDTO
	if err := ctx.ShouldBindJSON(&dto); err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	product, err := c.service.UpdateProduct(ctx.Request.Context(), id, dto)
	if err != nil {
		if errors.Is(err, service.ErrProductNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Product not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusOK, product)
}

// PATCH /api/products/:id/publish
func (c *ProductController) PublishProduct(ctx *gin.Context) {
	idStr := ctx.Param("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: "Invalid product ID"})
		return
	}

	product, err := c.service.PublishProduct(ctx.Request.Context(), id)
	if err != nil {
		if errors.Is(err, service.ErrProductNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Product not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusOK, product)
}

// DELETE /api/products/:id
func (c *ProductController) DeleteProduct(ctx *gin.Context) {
	idStr := ctx.Param("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: "Invalid product ID"})
		return
	}

	err = c.service.DeleteProduct(ctx.Request.Context(), id)
	if err != nil {
		if errors.Is(err, service.ErrProductNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Product not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.Status(http.StatusNoContent)
}

// POST /api/products/:id/images
func (c *ProductController) AddImages(ctx *gin.Context) {
	idStr := ctx.Param("id")
	productID, err := uuid.Parse(idStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: "Invalid product ID"})
		return
	}

	var dto types.AddImagesDTO
	if err := ctx.ShouldBindJSON(&dto); err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	err = c.service.AddProductImages(ctx.Request.Context(), productID, dto.Images)
	if err != nil {
		if errors.Is(err, service.ErrProductNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Product not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusCreated, types.APIResponse{Success: true, Message: "Images added successfully"})
}

// POST /api/products/:id/variants
func (c *ProductController) CreateVariant(ctx *gin.Context) {
	idStr := ctx.Param("id")
	productID, err := uuid.Parse(idStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: "Invalid product ID"})
		return
	}

	var dto types.CreateVariantDTO
	if err := ctx.ShouldBindJSON(&dto); err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: err.Error()})
		return
	}
	dto.ProductID = productID

	variant, err := c.service.CreateVariant(ctx.Request.Context(), dto)
	if err != nil {
		if errors.Is(err, service.ErrProductNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Product not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusCreated, variant)
}

// GET /api/products/variants/:variantId
func (c *ProductController) GetVariantByID(ctx *gin.Context) {
	variantIDStr := ctx.Param("variantId")
	variantID, err := uuid.Parse(variantIDStr)
	if err != nil {
		ctx.JSON(http.StatusBadRequest, types.APIResponse{Success: false, Error: "Invalid variant ID"})
		return
	}

	variant, err := c.service.GetVariantByID(ctx.Request.Context(), variantID)
	if err != nil {
		if errors.Is(err, service.ErrVariantNotFound) {
			ctx.JSON(http.StatusNotFound, types.APIResponse{Success: false, Error: "Variant not found"})
			return
		}
		ctx.JSON(http.StatusInternalServerError, types.APIResponse{Success: false, Error: err.Error()})
		return
	}

	ctx.JSON(http.StatusOK, variant)
}
